File: frontend/src/stores/exceptions.ts
import { create } from 'zustand';
import { devtools, persist } from 'zustand/middleware';
import { apiClient } from '@/api/client';

// Types
export type ExceptionSeverity = 'critical' | 'high' | 'medium' | 'low';
export type ExceptionCategory = 'andon' | 'quote' | 'production' | 'quality' | 'a3' | 'obeya' | 'task' | 'training';
export type ExceptionStatus = 'open' | 'acknowledged' | 'in_progress' | 'resolved' | 'escalated';

export interface Exception {
  id: string;
  title: string;
  category: ExceptionCategory;
  severity: ExceptionSeverity;
  status: ExceptionStatus;
  created_at: string;
  due_date: string;
  owner: string;
  owner_id: string;
  department: string;
  description: string;
  resolution_time?: number; // minutes
  resolved_at?: string;
  escalated_at?: string;
  escalated_to?: string;
  blocked_reason?: string;
  related_entity_id?: string;
  related_entity_type?: string;
  tags: string[];
  comments_count: number;
  attachments_count: number;
}

export interface ExceptionTrend {
  period: string;
  date: string;
  critical: number;
  high: number;
  medium: number;
  low: number;
  resolved: number;
  created: number;
}

export interface ExceptionStats {
  total_open: number;
  critical_count: number;
  overdue_count: number;
  escalated_count: number;
  blocked_count: number;
  avg_resolution_time_minutes: number;
  resolved_today: number;
  created_today: number;
  by_category: Record<ExceptionCategory, number>;
  by_severity: Record<ExceptionSeverity, number>;
  by_status: Record<ExceptionStatus, number>;
}

interface ExceptionsState {
  // Data
  exceptions: Exception[];
  trends: ExceptionTrend[];
  stats: ExceptionStats | null;
  
  // Loading & Error States
  isLoading: boolean;
  error: string | null;
  lastFetchedAt: number | null;
  
  // Actions
  fetchExceptions: (filters?: {
    category?: ExceptionCategory;
    severity?: ExceptionSeverity;
    status?: ExceptionStatus;
  }) => Promise<void>;
  fetchExceptionById: (id: string) => Promise<Exception>;
  acknowledgeException: (id: string) => Promise<void>;
  escalateException: (id: string, escalateTo: string, reason: string) => Promise<void>;
  resolveException: (id: string, resolutionNotes: string) => Promise<void>;
  assignException: (id: string, ownerId: string) => Promise<void>;
  addComment: (id: string, comment: string) => Promise<void>;
  
  // Trends & Stats
  fetchTrends: (days?: number) => Promise<void>;
  fetchStats: () => Promise<void>;
  
  // Utility
  clearError: () => void;
}

const CACHE_DURATION = 30000; // 30 seconds
const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000/api/v1';

export const useExceptionsStore = create<ExceptionsState>()(
  devtools(
    persist(
      (set, get) => ({
        // Initial State
        exceptions: [],
        trends: [],
        stats: null,
        isLoading: false,
        error: null,
        lastFetchedAt: null,
        
        // Actions
        fetchExceptions: async (filters = {}) => {
          const now = Date.now();
          const { lastFetchedAt, isLoading } = get();

          if (isLoading) {
            return;
          }
          
          // Use cache if fresh and no filters
          if (!Object.keys(filters).length && lastFetchedAt && now - lastFetchedAt < CACHE_DURATION) {
            return;
          }
          
          set({ isLoading: true, error: null });
          
          try {
            const params = new URLSearchParams(filters as any);
            const data = await apiClient.get<any>(`/exceptions?${params}`);

            const items = Array.isArray(data?.items)
              ? data.items
              : Array.isArray(data)
                ? data
                : [];
            
            set({
              exceptions: items,
              isLoading: false,
              lastFetchedAt: now,
            });
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch exceptions',
              isLoading: false,
            });
          }
        },
        
        fetchExceptionById: async (id: string) => {
          set({ isLoading: true, error: null });
          
          try {
            const exception = await apiClient.get<Exception>(`/exceptions/${id}`);
            set({ isLoading: false });
            return exception;
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch exception',
              isLoading: false,
            });
            throw error;
          }
        },
        
        acknowledgeException: async (id: string) => {
          set({ isLoading: true, error: null });
          
          try {
            await apiClient.post(`/exceptions/${id}/acknowledge`);
            
            set((state) => ({
              exceptions: state.exceptions.map(exc =>
                exc.id === id ? { ...exc, status: 'acknowledged' as ExceptionStatus } : exc
              ),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to acknowledge exception',
              isLoading: false,
            });
            throw error;
          }
        },
        
        escalateException: async (id: string, escalateTo: string, reason: string) => {
          set({ isLoading: true, error: null });
          
          try {
            await apiClient.post(`/exceptions/${id}/escalate`, { escalate_to: escalateTo, reason });
            
            set((state) => ({
              exceptions: state.exceptions.map(exc =>
                exc.id === id
                  ? {
                      ...exc,
                      status: 'escalated' as ExceptionStatus,
                      escalated_at: new Date().toISOString(),
                      escalated_to: escalateTo,
                    }
                  : exc
              ),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to escalate exception',
              isLoading: false,
            });
            throw error;
          }
        },
        
        resolveException: async (id: string, resolutionNotes: string) => {
          set({ isLoading: true, error: null });
          
          try {
            await apiClient.post(`/exceptions/${id}/resolve`, { resolution_notes: resolutionNotes });
            
            const resolvedAt = new Date().toISOString();
            const exception = get().exceptions.find(e => e.id === id);
            const resolutionTime = exception
              ? Math.floor((new Date(resolvedAt).getTime() - new Date(exception.created_at).getTime()) / 60000)
              : 0;
            
            set((state) => ({
              exceptions: state.exceptions.map(exc =>
                exc.id === id
                  ? {
                      ...exc,
                      status: 'resolved' as ExceptionStatus,
                      resolved_at: resolvedAt,
                      resolution_time: resolutionTime,
                    }
                  : exc
              ),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to resolve exception',
              isLoading: false,
            });
            throw error;
          }
        },
        
        assignException: async (id: string, ownerId: string) => {
          set({ isLoading: true, error: null });
          
          try {
            await apiClient.post(`/exceptions/${id}/assign`, { owner_id: ownerId });
            
            set((state) => ({
              exceptions: state.exceptions.map(exc =>
                exc.id === id ? { ...exc, owner_id: ownerId } : exc
              ),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to assign exception',
              isLoading: false,
            });
            throw error;
          }
        },
        
        addComment: async (id: string, comment: string) => {
          set({ isLoading: true, error: null });
          
          try {
            await apiClient.post(`/exceptions/${id}/comments`, { comment });
            
            set((state) => ({
              exceptions: state.exceptions.map(exc =>
                exc.id === id
                  ? { ...exc, comments_count: exc.comments_count + 1 }
                  : exc
              ),
              isLoading: false,
            }));
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to add comment',
              isLoading: false,
            });
            throw error;
          }
        },
        
        fetchTrends: async (days = 7) => {
          set({ isLoading: true, error: null });
          
          try {
            const data = await apiClient.get<any>(`/exceptions/trends?days=${days}`);

            const trends = Array.isArray(data?.items)
              ? data.items
              : Array.isArray(data)
                ? data
                : [];

            set({ trends, isLoading: false });
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch trends',
              isLoading: false,
            });
          }
        },
        
        fetchStats: async () => {
          set({ isLoading: true, error: null });
          
          try {
            const data = await apiClient.get<any>('/exceptions/summary');
            set({ stats: data ?? null, isLoading: false });
          } catch (error) {
            set({
              error: error instanceof Error ? error.message : 'Failed to fetch stats',
              isLoading: false,
            });
          }
        },
        
        // Utility
        clearError: () => set({ error: null }),
      }),
      {
        name: 'exceptions-storage',
        partialize: (state) => ({
          exceptions: state.exceptions,
          trends: state.trends,
        }),
      }
    )
  )
);

File: sensei-rs/crates/sensei-frontend/src/stores/exceptions.rs
//! Exceptions (operational issues / escalations) reactive store.
//!
//! Mirrors the Zustand [`exceptions.ts`](frontend/src/stores/exceptions.ts) store.

use crate::api::client::{ApiClient, ApiError};
use leptos::prelude::*;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// An operational exception.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExceptionDto {
    pub id: String,
    pub title: String,
    pub description: String,
    pub severity: String,
    pub status: String,
    pub category: Option<String>,
    pub source: Option<String>,
    pub owner: Option<String>,
    pub assigned_to: Option<String>,
    pub escalated_to: Option<String>,
    pub resolution_notes: Option<String>,
    pub created_at: String,
    pub updated_at: Option<String>,
}

/// Exception trend data.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExceptionTrendDto {
    pub date: String,
    pub count: i32,
    pub by_severity: HashMap<String, i32>,
    pub by_category: HashMap<String, i32>,
}

/// Exception statistics.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExceptionStatsDto {
    pub total: i32,
    pub open: i32,
    pub acknowledged: i32,
    pub escalated: i32,
    pub resolved: i32,
    pub avg_resolution_time_hours: f64,
}

/// Reactive store for exceptions.
#[derive(Debug, Clone)]
pub struct ExceptionsStore {
    pub exceptions: RwSignal<Vec<ExceptionDto>>,
    pub trends: RwSignal<Vec<ExceptionTrendDto>>,
    pub stats: RwSignal<Option<ExceptionStatsDto>>,
    pub loading: RwSignal<bool>,
    pub error: RwSignal<Option<String>>,
}

impl ExceptionsStore {
    pub fn new() -> Self {
        Self {
            exceptions: RwSignal::new(Vec::new()),
            trends: RwSignal::new(Vec::new()),
            stats: RwSignal::new(None),
            loading: RwSignal::new(false),
            error: RwSignal::new(None),
        }
    }

    /// Fetch exceptions with optional filters.
    pub async fn fetch_exceptions(&self, client: &ApiClient, filters: &HashMap<String, String>) {
        self.loading.set(true);
        self.error.set(None);
        let mut query = Vec::new();
        for (k, v) in filters {
            query.push(format!("{}={}", k, v));
        }
        let path = if query.is_empty() {
            "/api/v1/exceptions".to_string()
        } else {
            format!("/api/v1/exceptions?{}", query.join("&"))
        };
        match client.get::<Vec<ExceptionDto>>(&path).await {
            Ok(data) => self.exceptions.set(data),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.loading.set(false);
    }

    /// Fetch a single exception by ID.
    pub async fn fetch_exception_by_id(&self, client: &ApiClient, id: &str) {
        self.loading.set(true);
        self.error.set(None);
        match client
            .get::<ExceptionDto>(&format!("/api/v1/exceptions/{}", id))
            .await
        {
            Ok(data) => {
                self.exceptions.update(|excs| {
                    if let Some(pos) = excs.iter().position(|x| x.id == id) {
                        excs[pos] = data;
                    } else {
                        excs.push(data);
                    }
                });
            }
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.loading.set(false);
    }

    /// Acknowledge an exception.
    pub async fn acknowledge_exception(
        &self,
        client: &ApiClient,
        id: &str,
    ) -> Result<ExceptionDto, ApiError> {
        let exc: ExceptionDto = client
            .post(
                &format!("/api/v1/exceptions/{}/acknowledge", id),
                &serde_json::json!({}),
            )
            .await?;
        self.exceptions.update(|excs| {
            if let Some(pos) = excs.iter().position(|x| x.id == id) {
                excs[pos] = exc.clone();
            }
        });
        Ok(exc)
    }

    /// Escalate an exception.
    pub async fn escalate_exception(
        &self,
        client: &ApiClient,
        id: &str,
        escalate_to: &str,
        reason: &str,
    ) -> Result<ExceptionDto, ApiError> {
        let payload = serde_json::json!({
            "escalated_to": escalate_to,
            "reason": reason,
        });
        let exc: ExceptionDto = client
            .post(&format!("/api/v1/exceptions/{}/escalate", id), &payload)
            .await?;
        self.exceptions.update(|excs| {
            if let Some(pos) = excs.iter().position(|x| x.id == id) {
                excs[pos] = exc.clone();
            }
        });
        Ok(exc)
    }

    /// Resolve an exception.
    pub async fn resolve_exception(
        &self,
        client: &ApiClient,
        id: &str,
        resolution_notes: &str,
    ) -> Result<ExceptionDto, ApiError> {
        let payload = serde_json::json!({ "resolution_notes": resolution_notes });
        let exc: ExceptionDto = client
            .post(&format!("/api/v1/exceptions/{}/resolve", id), &payload)
            .await?;
        self.exceptions.update(|excs| {
            if let Some(pos) = excs.iter().position(|x| x.id == id) {
                excs[pos] = exc.clone();
            }
        });
        Ok(exc)
    }

    /// Assign an exception to an owner.
    pub async fn assign_exception(
        &self,
        client: &ApiClient,
        id: &str,
        owner_id: &str,
    ) -> Result<ExceptionDto, ApiError> {
        let payload = serde_json::json!({ "owner_id": owner_id });
        let exc: ExceptionDto = client
            .post(&format!("/api/v1/exceptions/{}/assign", id), &payload)
            .await?;
        self.exceptions.update(|excs| {
            if let Some(pos) = excs.iter().position(|x| x.id == id) {
                excs[pos] = exc.clone();
            }
        });
        Ok(exc)
    }

    /// Add a comment to an exception.
    pub async fn add_comment(
        &self,
        client: &ApiClient,
        id: &str,
        comment: &str,
    ) -> Result<serde_json::Value, ApiError> {
        let payload = serde_json::json!({ "comment": comment });
        client
            .post(&format!("/api/v1/exceptions/{}/comments", id), &payload)
            .await
    }

    /// Fetch exception trends.
    pub async fn fetch_trends(&self, client: &ApiClient, days: i32) {
        self.loading.set(true);
        self.error.set(None);
        match client
            .get::<Vec<ExceptionTrendDto>>(&format!("/api/v1/exceptions/trends?days={}", days))
            .await
        {
            Ok(data) => self.trends.set(data),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.loading.set(false);
    }

    /// Fetch exception stats.
    pub async fn fetch_stats(&self, client: &ApiClient) {
        self.loading.set(true);
        self.error.set(None);
        match client
            .get::<ExceptionStatsDto>("/api/v1/exceptions/stats")
            .await
        {
            Ok(data) => self.stats.set(Some(data)),
            Err(e) => self.error.set(Some(e.to_string())),
        }
        self.loading.set(false);
    }
}

impl Default for ExceptionsStore {
    fn default() -> Self {
        Self::new()
    }
}
